crypto/fipsmodule/service_indicator/service_indicator.c

```diff
@@ -225,23 +225,34 @@ static void evp_md_ctx_verify_service_indicator(const EVP_MD_CTX *ctx,
     if (!EVP_PKEY_CTX_get_rsa_padding(pctx, &padding)) {
       goto err;
     }
+    // The approved RSA key sizes for signing are 2048, 3072 and 4096 bits.
+    // Note: |EVP_PKEY_size| returns the size in bytes.
+    size_t pkey_size = EVP_PKEY_size(ctx->pctx->pkey);
+
     if (padding == RSA_PKCS1_PSS_PADDING) {
       int salt_len;
+      int hash_len = (int)EVP_MD_size(pctx_md);
       const EVP_MD *mgf1_md;
       if (!EVP_PKEY_CTX_get_rsa_pss_saltlen(pctx, &salt_len) ||
           !EVP_PKEY_CTX_get_rsa_mgf1_md(pctx, &mgf1_md) ||
-          (salt_len != -1 && salt_len != (int)EVP_MD_size(pctx_md)) ||
           EVP_MD_type(mgf1_md) != md_type) {
-        // Only PSS where saltLen == hashLen is tested with ACVP. Cases with
-        // non-standard padding functions are also excluded.
+        goto err;
+      }
+      // RSA_PSS_SALTLEN_DIGEST (-1) has the special meaning of "sLen ==
+      // hLen", other values need additional checks.
+      // FIPS 186-4, Section 5.5, requires that if the RSA key size is 1024
```

Comment:
but 1024 is not an allowed length, so no need to consider it?

And the stronger check above `salt_len != (int)EVP_MD_size(pctx_md)` is already sufficient?

Reply:
> but 1024 is not an allowed length, so no need to consider it?

Sure, I think we can remove the check for that, since it's not doable.

> And the stronger check above salt_len != (int)EVP_MD_size(pctx_md) is already sufficient?

The original comment implied that the original check was implemented since ACVP test cases only had `saltLen == hashLen` for PSS. Outside of ACVP test cases, the FIPS guidance states that `0 <= salt_len <= hash_len` is usable for FIPS as well, so I thought we should be marking these cases as APPROVED.